sygnal/helper/proxy/proxy_asyncio.py

```diff
@@ -0,0 +1,241 @@
+
+        logger.debug("Initiating proxy CONNECT")
+
+        # now we wait ...
+        self.transport = transport
+
+
+class ProxyingEventLoopWrapper:
+    """
+    This is a wrapper for an asyncio.AbstractEventLoop which intercepts calls to
+    create_connection and transparently tunnels them through an HTTP CONNECT
+    proxy.
+    """
+
+    def __init__(
+        self, wrapped_loop: asyncio.AbstractEventLoop, proxy_url_str: str,
+    ):
+        """
+        Args:
+            wrapped_loop:
+                the underlying Event Loop to wrap
+            proxy_url_str (str):
+                The address of the HTTP proxy to use.
+                Used to connect to the proxy, as well as in the `Host` request
+                header to the proxy, and for the extraction of basic
+                authentication credentials (if required).
+
+                Examples: 'http://127.0.3.200:8080'
+                       or 'http://user:secret@prox:8080'
+        """
+        self._wrapped_loop = wrapped_loop
+        self.proxy_url_str = proxy_url_str
+
+    async def create_connection(
+        self,
+        protocol_factory: Callable[[], asyncio.Protocol],
+        host: str,
+        port: int,
+        ssl: Union[bool, SSLContext] = False,
+    ):
+        proxy_url_parts = decompose_http_proxy_url(self.proxy_url_str)
+
+        def make_protocol():
+            proxy_setup_protocol = HttpConnectProtocol(
+                proxy_url_parts, f"{host}:{port}"
+            )
+            return proxy_setup_protocol
+
+        # create a raw TCP connection to the proxy
+        # (N.B. if we want to ever use TLS to the proxy [e.g. to protect the proxy
+        # credentials], we can ask this to give us a TLS connection).
+        transport, connect_protocol = await self._wrapped_loop.create_connection(
+            make_protocol, proxy_url_parts.hostname, proxy_url_parts.port
+        )
+
+        assert isinstance(connect_protocol, HttpConnectProtocol)
+
+        # wait for the HTTP Proxy CONNECT sequence to complete
+        transport = await connect_protocol.wait_for_establishment
+
+        if ssl:
+            if isinstance(ssl, SSLContext):
+                sslcontext = ssl
+            else:
+                sslcontext = SSLContext()
+
+            # be careful not to use the `transport` ever again after passing it
+            # to start_tls — we overwrite our variable with the TLS-wrapped
+            # transport to avoid that!
+            transport = await self._wrapped_loop.start_tls(
+                transport, connect_protocol, sslcontext
+            )
+
+        # construct the desired protocol and hand over the transport to it
+        new_protocol = protocol_factory()
+        # wire up transport to call `data_received` etc. on the new transport
+        transport.set_protocol(new_protocol)
+        # let the protocol know it has been connected to the transport
+        new_protocol.connection_made(transport)
```

Comment:
slightly wondering if it would be better to pass `sslcontext` and `protocol_factory` into HttpConnectProtocol's constructor, and moving this logic inside `HttpConnectProtocol.data_received`. It might make handling dangling data in the `buffer` easier, and would give you a CONNECT implementation that can be used without having to wrap the event loop, if that is ever useful.

Reply:
Yes, I agree, but `loop.start_tls(` is an `async def`, whereas none of the stuff in `Protocol` is `async`, it's all callback driven.

Reply:
I think I have a plan